crates/nargo/src/ops/foreign_calls.rs

```diff
@@ -63,31 +64,89 @@ impl ForeignCall {
+            convert_fmt_string_inputs(foreign_call_inputs)?
+        } else {
+            convert_string_inputs(foreign_call_inputs)?
+        };
+        println!("{output_string}");
         Ok(())
     }
 }
 
-/// Fetch the abi type from the foreign call input
-/// The remaining input values should hold the values to be printed
-fn fetch_abi_type(
-    foreign_call_inputs: &[Vec<Value>],
-) -> Result<(AbiType, &[Vec<Value>]), ForeignCallError> {
+fn convert_string_inputs(foreign_call_inputs: &[Vec<Value>]) -> Result<String, ForeignCallError> {
+    // Fetch the abi type from the foreign call input
+    // The remaining input values should hold what is to be printed
     let (abi_type_as_values, input_values) =
         foreign_call_inputs.split_last().ok_or(ForeignCallError::MissingForeignCallInputs)?;
+    let abi_type = fetch_abi_type(abi_type_as_values)?;
+
+    // We must use a flat map here as each value in a struct will be in a separate input value
+    let mut input_values_as_fields =
+        input_values.iter().flat_map(|values| vecmap(values, |value| value.to_field()));
+
+    let decoded_value = decode_value(&mut input_values_as_fields, &abi_type)?;
+    let json_value = JsonTypes::try_from_input_value(&decoded_value, &abi_type)?;
+
+    Ok(json_value.to_string())
+}
+
+fn convert_fmt_string_inputs(
+    foreign_call_inputs: &[Vec<Value>],
+) -> Result<String, ForeignCallError> {
+    let (message_as_values, input_and_abi_values) =
+        foreign_call_inputs.split_first().ok_or(ForeignCallError::MissingForeignCallInputs)?;
+
+    let message_as_fields = vecmap(message_as_values, |value| value.to_field());
+    let message_as_string = decode_string_value(&message_as_fields);
+
+    let (num_values, input_and_abi_values) =
+        input_and_abi_values.split_first().ok_or(ForeignCallError::MissingForeignCallInputs)?;
+
+    let mut output_strings = Vec::new();
+    let num_values = num_values[0].to_field().to_u128() as usize;
+
+    let mut abi_types = Vec::new();
+    for abi_values in input_and_abi_values.iter().skip(input_and_abi_values.len() - num_values) {
+        let abi_type = fetch_abi_type(abi_values)?;
+        abi_types.push(abi_type);
+    }
+
+    for i in 0..num_values {
+        let abi_type = &abi_types[i];
+        let type_size = abi_type.field_count() as usize;
+
+        let mut input_values_as_fields = input_and_abi_values[i..(i + type_size)]
+            .iter()
+            .flat_map(|values| vecmap(values, |value| value.to_field()));
+
+        let decoded_value = decode_value(&mut input_values_as_fields, abi_type)?;
+
+        let json_value = JsonTypes::try_from_input_value(&decoded_value, abi_type)?;
+        output_strings.push(json_value.to_string());
+    }
+
+    let mut output_strings_iter = output_strings.into_iter();
+    let re = Regex::new(r"\{([a-zA-Z0-9]+)\}")
+        .expect("ICE: an invalid regex pattern was used for checking format strings");
```

Comment:
I'd be wary of regexes since they tend to be greedy by default

Reply:
I thought this would be fine as I'm ok being greedy in this case. I also am resolving the fmt string using the regex so the idents inside of the fmt string should all be in constructed in a greedy manner as well. Were you thinking something more like directly checking for curly braces in the string?